04.ls/myls.rb

```diff
@@ -0,0 +1,138 @@
+  {
+    sorted_directories: [],
+    empty_directories: [],
+    error_messages: [],
+    file_results: []
+  }
+end
+
+def file_type(filenames)
+  if File.directory?(filenames)
+    :directory_path
+  elsif File.file?(filenames)
+    :file_path
+  else
+    :invalid
+  end
+end
+
+def create_sorted_directory(filenames)
+  directory_files = Dir.glob("#{filenames}/*").map { File.basename(_1) }
+  directory_name = filenames
+
+  row_count = calculate_row_count(directory_files)
+  formatted_data = create_formatted_data(directory_files, row_count)
+
+  {
+    directory_name:,
+    row_count:,
+    formatted_data:
+  }
+end
+
+def create_error_message(filenames)
+  "ls: '#{filenames}' にアクセスできません：そのようなファイルやディレクトリはありません"
+end
+
+def calculate_row_count(directory_files)
+  row_count = (directory_files.size + DISPLAY_COLUMNS_COUNT - 1) / DISPLAY_COLUMNS_COUNT
+  row_count.zero? ? 1 : row_count
+end
+
+def create_formatted_data(directory_files, row_count)
+  formatted_data = directory_files.each_slice(row_count).to_a
+  formatted_data << [''] while formatted_data.size < DISPLAY_COLUMNS_COUNT
+  formatted_data
+end
+
+def result_display(result_data, multiple_arguments_received)
+  display_error_messages(result_data[:error_messages]) unless result_data[:error_messages].empty?
+  unless result_data[:file_results].empty?
+    need_additional_line_break = if result_data[:sorted_directories].empty?
+                                   false
+                                 else
+                                   true
+                                 end
+    display_file_results(result_data[:file_results], need_additional_line_break)
+  end
+  display_directory_results(result_data[:sorted_directories], multiple_arguments_received) unless result_data[:sorted_directories].empty?
+end
+
+def display_error_messages(error_messages)
+  error_messages.each { puts _1 }
+end
+
+def display_file_results(file_results, need_additional_line_break)
+  puts file_results.sort.join('  ')
+  puts if need_additional_line_break
+end
+
+def display_directory_results(sorted_directories, multiple_arguments_received)
+  sorted_directories.each.with_index do |directory_data, directory_number|
+    display_file_name(directory_data) if multiple_arguments_received
+    max_column_widths = calculate_max_column_widths(directory_data[:formatted_data])
+    directory_data[:row_count].times do |row|
+      DISPLAY_COLUMNS_COUNT.times do |col|
+        target_data = directory_data[:formatted_data][col][row]
+        wide_chars_count = count_multibyte_characters(target_data) || 0
+        print target_data.to_s.ljust((max_column_widths[col] || 0) + 2 - wide_chars_count)
+      end
+      puts
+    end
+    puts if sorted_directories.size - 1 != directory_number
+  end
+end
+
+def display_file_name(directory_data)
+  if directory_data[:formatted_data] == [[''], [''], ['']]
+    print "#{directory_data[:directory_name]}:"
+  else
+    puts "#{directory_data[:directory_name]}:"
+  end
```

Comment:
条件分岐は局所的なほうがわかりやすいです。

```suggestion
  print "#{directory_data[:directory_name]}:"
  puts unless directory_data[:formatted_data].all?([''])
```

Reply:
これも引数がスタンプ結合にならないようにして、

```ruby
  print "#{directory_name}:"
  puts if has_any_file
```

のようにできるといいかもしれませんね。

Reply:
スタンプ結合を出来るだけ回避できるように修正してみました